tests/test_flow/fakes.py

```diff
@@ -171,28 +117,65 @@ def pipe(self, url, content_bytes):
             f.write(content_bytes)
 
 
-def fake_aip_client(gcs_fs, caplog):
-    def create_aip_job(body, parent):
-        # AIP executions do not transmit logs to local instance of bionic.
-        # Emulate this behavior by setting the log level; this should filter out
-        # all (or almost all) the logs. Tests that check the log output will
-        # only test against logs which are printed locally.
-        with caplog.at_level(logging.CRITICAL):
-            assert body["trainingInput"]["args"][:3] == [
-                "python",
-                "-m",
-                "bionic.aip.main",
-            ]
-            run_aip(body["trainingInput"]["args"][3], gcs_fs)
-        return Mock()
+@contextmanager
+def temporarily_filter_all_logs():
+    class FilterAllLogs(logging.Filter):
+        def filter(self, rec):
+            return 0
+
+    logger = logging.getLogger()
+    filter_all_logs = FilterAllLogs()
+    try:
+        logger.addFilter(filter_all_logs)
+        yield
+    finally:
+        logger.removeFilter(filter_all_logs)
+
+
+@contextmanager
+def temporarily_clean_cache_dir(tmp_path):
+    cache_dir = str(tmp_path / "BNTESTDATA")
+    assert os.path.exists(cache_dir)
 
-    mock_aip_client = Mock()
-    mock_aip_client.projects().jobs().create = create_aip_job
-    mock_aip_client.projects().jobs().get().execute.return_value = {
-        "state": "SUCCEEDED"
-    }
+    backup_dir = str(tmp_path / "BACKUP")
+    assert not os.path.exists(backup_dir)
+
+    try:
+        os.rename(cache_dir, backup_dir)
+        yield
+    finally:
+        shutil.rmtree(cache_dir)
+        os.rename(backup_dir, cache_dir)
+
+
+class FakeAipClient:
+    def __init__(self, gcs_fs, tmp_path):
+        self.gcs_fs = gcs_fs
+        self.tmp_path = tmp_path
+
+    def _create_aip_job(self, body, parent):
+        assert body["trainingInput"]["args"][:3] == [
+            "python",
+            "-m",
+            "bionic.aip.main",
+        ]
+
+        input_uri = body["trainingInput"]["args"][3]
+        assert self.gcs_fs.exists(input_uri)
+
+        # AIP execution do not have access to local disk cache and do not send
+        # back logs to local bionic instance. Emulate this behavior while
+        # running the AIP job.
+        with temporarily_filter_all_logs(), temporarily_clean_cache_dir(self.tmp_path):
```

Comment:
It looks like we just run this job synchronously. Does this mean it's impossible for us to have a test case where multiple fake AIP jobs are in progress at once?

Reply:
We are calling `executor.submit` one at a time. So I think if we are running the job synchronously during creation, we should not have multiple fake AIP jobs running concurrently. But as you mentioned, it may be possible for a parallel executor to be running at the same time.